File: src/org/smallbean/interview/VideoCapture.java
package org.smallbean.interview;

import java.io.IOException;

import org.smallbean.interview.utilities.Data;

import android.app.Activity;
import android.content.Context;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.util.Log;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.Window;
import android.view.WindowManager;
import android.media.MediaRecorder;
import android.view.View;
import android.view.SurfaceHolder.Callback;
import android.view.View.OnClickListener;
import android.widget.Toast;

public class VideoCapture extends Activity implements OnClickListener, Callback 
{
	private String imageFilePath;
	
    private boolean recording = false;
	private MediaRecorder recorder;
	private SurfaceHolder holder;

    private Data data = Data.getInstance();
	
	@Override
    public void onCreate(Bundle savedInstanceState) 
	{
        super.onCreate(savedInstanceState);
        
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
        
        setContentView(R.layout.cameraview);
        
        imageFilePath = data.GetNewVideoURL();
        Log.d("VideoPath", imageFilePath);
        
        
        initRecorder();
        SurfaceView cameraView = (SurfaceView) findViewById(R.id.surface);
        holder = cameraView.getHolder();
        holder.addCallback(this);
        holder.setType(SurfaceHolder.SURFACE_TYPE_PUSH_BUFFERS);

        cameraView.setClickable(true);
        cameraView.setOnClickListener(this);
         
    }

	private void initRecorder()
	{
		recorder = new MediaRecorder();
		
		recorder.setAudioSource(MediaRecorder.AudioSource.DEFAULT);
		recorder.setVideoSource(MediaRecorder.VideoSource.DEFAULT);
		
		recorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
		recorder.setAudioEncoder(MediaRecorder.AudioEncoder.DEFAULT);
		recorder.setVideoEncoder(MediaRecorder.VideoEncoder.DEFAULT);
		recorder.setOutputFile(imageFilePath);
		
		recorder.setMaxDuration(50000);
		recorder.setMaxFileSize(5000000);
	}
	
	@Override
	public void onClick(View v)
	{
		if(recording)
		{
			
			recorder.stop();
			showToast(this,"Recording stopped");
			recording = false;
			finish();
			//initRecorder();
			//prepareRecorder();
		}
		else
		{
			recording = true;
			recorder.start();
			
			showToast(this,"Recording started");
		}
	}

	@Override
	public void surfaceChanged(SurfaceHolder holder, int format, int width,
			int height) {
		// TODO Auto-generated method stub

	}

	private void prepareRecorder()
	{
		
		recorder.setPreviewDisplay(holder.getSurface());
		try
		{
			
			recorder.prepare();
		}
		catch(IllegalStateException e)
		{
			
			e.printStackTrace();
			finish();
		}
		catch(IOException e)
		{
		    e.printStackTrace();
		    finish();
		}
	}
	@Override
	public void surfaceCreated(SurfaceHolder holder) 
	{
		
		prepareRecorder();
	}

	@Override
	public void surfaceDestroyed(SurfaceHolder holder)
	{
		if(recording)
		{
			recorder.stop();
			recording = false;
		}
		recorder.release();
		finish();
	}
	
	

  	public void captureVideo(View view) 
  	{		    	    
          showToast(this,imageFilePath);       	               
                  
  		
  	}

  
  private void showToast(Context mContext, String text) {
  	Toast.makeText(mContext, text, Toast.LENGTH_SHORT).show();
  }

}

File: src/org/smallbean/interview/VideoCapture.scala
package org.smallbean.interview

import java.io.IOException

import org.smallbean.interview.utilities.Data

import android.app.Activity
import android.content.Context
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.util.Log
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.Window
import android.view.WindowManager
import android.media.MediaRecorder
import android.view.View
import android.view.SurfaceHolder.Callback
import android.view.View.OnClickListener
import android.widget.Toast


class VideoCapture extends Activity with OnClickListener with Callback 
{
    var recording = false
	var recorder:MediaRecorder = null
	var holder:SurfaceHolder = null
	var imageFilePath = ""
	
	override def onCreate(savedInstanceState:Bundle) 
	{
        super.onCreate(savedInstanceState)
        
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE)
        
        setContentView(R.layout.cameraview)
        
        imageFilePath = Data.GetNewVideoURL
        Log.d("VideoPath", imageFilePath)
        
        
        initRecorder
        val cameraView = findViewById(R.id.surface).asInstanceOf[SurfaceView]
        holder = cameraView.getHolder()
        holder.addCallback(this)
        holder.setType(SurfaceHolder.SURFACE_TYPE_PUSH_BUFFERS)

        cameraView.setClickable(true)
        cameraView.setOnClickListener(this)
         
    }

	private def initRecorder
	{
		recorder = new MediaRecorder()
		
		recorder.setAudioSource(recorder.AudioSource.DEFAULT)
		recorder.setVideoSource(recorder.VideoSource.DEFAULT)
		recorder.setOutputFormat(recorder.OutputFormat.MPEG_4)
		recorder.setAudioEncoder(recorder.AudioEncoder.DEFAULT)
		recorder.setVideoEncoder(recorder.VideoEncoder.DEFAULT)
		recorder.setOutputFile(imageFilePath)
		
		recorder.setMaxDuration(50000)
		recorder.setMaxFileSize(5000000)
	}
	
	override def onClick(v:View)
	{
		if(recording)
		{
			
			recorder.stop()
			showToast(this,"Recording stopped")
			recording = false
			finish()
		}
		else
		{
			recording = true
			recorder.start()
			
			showToast(this,"Recording started")
		}
	}

	override def surfaceChanged(holder:SurfaceHolder, format:Int, width:Int, height:Int) {
		// TODO Auto-generated method stub

	}

	private def prepareRecorder
	{
		
		recorder.setPreviewDisplay(holder.getSurface())
		try
		{
			
			recorder.prepare()
		}
		catch
		{
			case _ => finish()
		}
	}
	
	override def surfaceCreated(holder:SurfaceHolder) 
	{		
		prepareRecorder
	}

	override def surfaceDestroyed(holder:SurfaceHolder)
	{
		if(recording)
		{
			recorder.stop()
			recording = false
		}
		recorder.release()
		finish()
	}
	
	  
  private def showToast(mContext:Context, text:String) {
  	Toast.makeText(mContext, text, Toast.LENGTH_SHORT).show()
  }

}
